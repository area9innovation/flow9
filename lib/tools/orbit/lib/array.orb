// Array utility functions library for Orbit

// map: Apply a function to each element in an array
fn map(arr, f) = (
    if length(arr) = 0 then []
    else [f(arr[0])] + map(subrange(arr, 1, length(arr) - 1), f)
);

// filter: Keep only elements that match a predicate
fn filter(arr, pred) = (
    if length(arr) = 0 then []
    else if pred(arr[0]) then 
        [arr[0]] + filter(subrange(arr, 1, length(arr) - 1), pred)
    else 
        filter(subrange(arr, 1, length(arr) - 1), pred)
);

// fold: Accumulate a result based on each array element and an accumulator
fn fold(arr, init, f) = (
    if length(arr) = 0 then init
    else fold(subrange(arr, 1, length(arr) - 1), f(init, arr[0]), f)
);

// mapi: Map with index
fn mapi(arr, f) = (
    fn mapiHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else [f(arr[0], idx)] + mapiHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
    );
    mapiHelper(arr, f, 0)
);

// filteri: Filter with index
fn filteri(arr, pred) = (
    fn filteriHelper(arr, pred, idx) = (
        if length(arr) = 0 then []
        else if pred(arr[0], idx) then 
            [arr[0]] + filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
        else 
            filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
    );
    filteriHelper(arr, pred, 0)
);

// foldi: Fold with index
fn foldi(arr, init, f) = (
    fn foldiHelper(arr, acc, f, idx) = (
        if length(arr) = 0 then acc
        else foldiHelper(subrange(arr, 1, length(arr) - 1), f(acc, arr[0], idx), f, idx + 1)
    );
    foldiHelper(arr, init, f, 0)
);

// tail: Return all elements except the first one
fn tail(arr) = (
    if length(arr) <= 1 then []
    else subrange(arr, 1, length(arr) - 1)
);

// tailFrom: Return all elements from a given index to the end
fn tailFrom(arr, idx) = (
    if idx >= length(arr) then []
    else subrange(arr, idx, length(arr) - idx)
);

// take: Return the first n elements of an array
fn take(arr, n) = (
    if n <= 0 || length(arr) = 0 then []
    else [arr[0]] + take(subrange(arr, 1, length(arr) - 1), n - 1)
);

// iter: Call a function for each element (for side effects)
fn iter(arr, f) = (
    if length(arr) = 0 then 0 // Return 0 instead of ()
    else (
        f(arr[0]);
        iter(subrange(arr, 1, length(arr) - 1), f)
    )
);

// iteri: Iterate with index
fn iteri(arr, f) = (
    fn iteriHelper(arr, f, idx) = (
        if length(arr) = 0 then 0 // Return 0 instead of ()
        else (
            f(arr[0], idx);
            iteriHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
        )
    );
    iteriHelper(arr, f, 0)
);

// filtermap: Combine filter and map operations
fn filtermap(arr, f) = (
    if length(arr) = 0 then []
    else (
        let result = f(arr[0]);
        result is (
            Some(val) => [val] + filtermap(subrange(arr, 1, length(arr) - 1), f);
            None() => filtermap(subrange(arr, 1, length(arr) - 1), f);
            _ => filtermap(subrange(arr, 1, length(arr) - 1), f) // For safety
        )
    )
);

// filtermapi: Filtermap with index
fn filtermapi(arr, f) = (
    fn filtermapHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else (
            let result = f(arr[0], idx);
            result is (
                Some(val) => [val] + filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                None() => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                _ => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1) // For safety
            )
        )
    );
    filtermapHelper(arr, f, 0)
);

// removeIndex: Remove an element at a specific index
fn removeIndex(arr, idx) = (
    if idx < 0 || idx >= length(arr) then arr
    else take(arr, idx) + tailFrom(arr, idx + 1)
);

// removeRange: Remove a range of elements
fn removeRange(arr, start, count) = (
    if start < 0 || start >= length(arr) || count <= 0 then arr
    else take(arr, start) + tailFrom(arr, start + count)
);

// removeFirst: Remove the first occurrence of a specific element
fn removeFirst(arr, element) = (
    fn removeFirstHelper(before, after) = (
        if length(after) = 0 then before
        else if after[0] = element then before + subrange(after, 1, length(after) - 1)
        else removeFirstHelper(before + [after[0]], subrange(after, 1, length(after) - 1))
    );
    removeFirstHelper([], arr)
);

// removeAll: Remove all occurrences of a specific element
fn removeAll(arr, element) = (
    filter(arr, \x.x != element != 0)
);

// contains: Check if an element exists in the array
fn contains(arr, element) = (
    if length(arr) = 0 then false
    else if arr[0] = element then true
    else contains(subrange(arr, 1, length(arr) - 1), element)
);

// exists: Check if any element satisfies a predicate
fn exists(arr, pred) = (
    if length(arr) = 0 then false
    else if pred(arr[0]) then true
    else exists(subrange(arr, 1, length(arr) - 1), pred)
);

// forall: Check if all elements satisfy a predicate
fn forall(arr, pred) = (
    if length(arr) = 0 then true
    else if !pred(arr[0]) then false
    else forall(subrange(arr, 1, length(arr) - 1), pred)
);

// uniq: Remove duplicates from an array
fn uniq(arr) = (
    if length(arr) <= 1 then arr
    else (
        fn uniq_helper(remaining, result) = (
            if length(remaining) = 0 then result
            else (
                let current = remaining[0];
                let rest = tail(remaining);
                
                if contains(result, current) then
                    // Skip duplicate
                    uniq_helper(rest, result)
                else
                    // Keep this element
                    uniq_helper(rest, result + [current])
            )
        );
        uniq_helper(arr, [])
    )
);