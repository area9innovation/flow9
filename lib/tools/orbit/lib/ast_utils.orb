// AST utility functions for Orbit

// Create a properly nested AST node by folding over an array of arguments
// This function handles arrays of any length, whereas makeAst expects exactly 2 arguments
fn makeFoldedAst(op, args) = (
    let len = length(args);
    
    if len = 0 then (
        // Handle empty array case - return an appropriate default value
        // This depends on the operator, but we'll return a neutral element
        if op = "&&" then true
        else if op = "||" then false
        else args  // Return empty array as fallback
    ) else if len = 1 then (
        // For single element, just return it without creating an operation
        args[0]
    ) else if len = 2 then (
        // For exactly two elements, we can use makeAst directly
        makeAst(op, args)
    ) else (
        // For 3+ elements, fold to create a proper binary tree
        // Start with the first element and fold over the rest
        let initialAst = args[0];
        let remainingArgs = subrange(args, 1, len - 1);
        
        // Fold over the remaining arguments, combining each with the accumulator
        fold(remainingArgs, initialAst, \acc, element -> (
            makeAst(op, [acc, element])
        ))
    )
);