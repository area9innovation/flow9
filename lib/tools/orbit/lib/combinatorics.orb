// Basic combinatorial functions
fn factorial(n) = (
	if n <= 1 then 1
	else n * factorial(n - 1)
);

fn binomial_coefficient(n, k) = (
	if k < 0 || k > n then 0
	else if k == 0 || k == n then 1
	else (
		// Use multiplicative formula to avoid large intermediate values
		fn binom_helper(i, result) = (
			if i > k then result
			else binom_helper(i + 1, result * (n - k + i) / i)
		);
		binom_helper(1, 1)
	)
);

fn permutations(n, r) = (
	if r < 0 || r > n then 0
	else if r == 0 then 1
	else (
		// Calculate n! / (n-r)!
		fold(range(n-r+1, n+1), 1, \product, i.product * i)
	)
);

// Generate permutations of an array
fn generate_permutations(arr) = (
	if length(arr) <= 1 then [arr]
	else (
		let result = [];
		fn perms_helper(i, perms) = (
			if i >= length(arr) then perms
			else (
				let element = arr[i];
				let rest = remove_at_index(arr, i);
				let sub_perms = generate_permutations(rest);

				let new_perms = fold(sub_perms, perms, \acc, p.acc + [[element] + p]);
				perms_helper(i + 1, new_perms)
			)
		);
		perms_helper(0, [])
	)
);

// Generate combinations of an array
fn generate_combinations(arr, r) = (
	if r > length(arr) then []
	else if r == 0 then [[]]
	else if r == length(arr) then [arr]
	else (
		if length(arr) == 0 then []
		else (
			let first = arr[0];
			let rest = subrange(arr, 1, length(arr) - 1);

			// Combinations that include the first element
			let with_first = map(generate_combinations(rest, r - 1),
													\comb.[first] + comb);

			// Combinations that exclude the first element
			let without_first = generate_combinations(rest, r);

			with_first + without_first
		)
	)
);
