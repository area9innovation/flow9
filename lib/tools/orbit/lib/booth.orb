import lib/array;

// Linear time Booth's algorithm to find canonical form in cyclic groups
fn booth_canonical(arr) = (
	if length(arr) <= 1 then arr
	else (
		let n = length(arr);

		// Create an auxiliary array which is the original array concatenated with itself
		let double_arr = arr + arr;

		// Initialize failure function array with -1
		// (We use -1 instead of conventional 0 to simplify logic)
		let f = \i. if i < 0 || i >= n then -1 else f(i);

		// Compute the min rotation index in linear time
		fn compute_min_index() = (
			// Initialize starting values
			fn find_min(i, j, k) = (
				// We're done when we've checked all rotations
				if i + k >= 2 * n then j
				else if k >= n then find_min(i + 1, j, 0)
				else (
					// Compute the current indices we're comparing
					let i_idx = (i + k) % (2 * n);
					let j_idx = (j + k) % (2 * n);

					if double_arr[i_idx] < double_arr[j_idx] then
						// i rotation is smaller, j becomes i
						find_min(i, i, k + 1)
					else if double_arr[i_idx] > double_arr[j_idx] then
						// j rotation is smaller, i moves forward
						find_min(i + k + 1, j, 0)
					else
						// Equal so far, continue comparison
						find_min(i, j, k + 1)
				)
			);

			// Start with first two positions
			find_min(0, 0, 0)
		);

		let min_start = compute_min_index();

		// Build the canonical form starting from min_start
		fn build_result(idx, result) = (
			if idx >= n then result
			else build_result(idx + 1, result + [arr[(min_start + idx) % n]])
		);

		build_result(0, [])
	)
);

