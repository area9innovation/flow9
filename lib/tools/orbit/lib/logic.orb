import lib/rewrite;
import lib/array;
import lib/reflect;
import lib/sort;

// Helper function to remove duplicates from an array
fn unique(arr) = (
    fold(arr, [], \acc, item -> (
        if contains(acc, item) then acc else concat(acc, [item])
    ))
);

// Helper function to check if an array is sorted
fn isSorted(arr) = (
    if length(arr) <= 1 then true
    else (
        forall(enumFromTo(0, length(arr) - 2), \i -> (
            arr[i] <= arr[i + 1]
        ))
    )
);

/**
 * Canonical Logic Forms: CNF and DNF implementation using automatic flattening
 * 
 * This library demonstrates how to implement Conjunctive Normal Form (CNF)
 * and Disjunctive Normal Form (DNF) in Orbit using automatic SExpr flattening
 * where expressions like a||b||c automatically become `||`(a,b,c) during lowering.
 */

// ----- REWRITE RULES -----

// Convert an expression to normal form by applying rewrite rules
fn to_normal_form(expr, is_cnf) = (
    // Define the basic logical rules
    let basic_rules = [
        // Double negation elimination
        ["double_negation", quote(!!a), quote(a), true],
        
        // De Morgan's laws
        ["de_morgan_and", quote(!(a && b)), quote(!a || !b), true],
        ["de_morgan_or", quote(!(a || b)), quote(!a && !b), true],
        
        // Contradiction and tautology laws
        ["contradiction", quote(a && !a), quote(false), true],
        ["tautology", quote(a || !a), quote(true), true],
        
        // Identity laws - with false
        ["or_with_false", quote(false || a), quote(a), true],
        ["or_with_false2", quote(a || false), quote(a), true],
        ["and_with_false", quote(a && false), quote(false), true],
        ["and_with_false2", quote(false && a), quote(false), true],
        
        // Identity laws - with true
        ["and_with_true", quote(true && a), quote(a), true],
        ["and_with_true2", quote(a && true), quote(a), true],
        ["or_with_true", quote(true || a), quote(true), true],
        ["or_with_true2", quote(a || true), quote(true), true],
        
        // Idempotent laws
        ["and_idempotent", quote(a && a), quote(a), true],
        ["or_idempotent", quote(a || a), quote(a), true],
        
        // Absorption laws - all variants
        ["absorption1", quote(a || (a && b)), quote(a), true],
        ["absorption2", quote((a && b) || a), quote(a), true],
        ["absorption3", quote(a && (a || b)), quote(a), true],
        ["absorption4", quote((a || b) && a), quote(a), true],
        
        // Exclusive-or simplification patterns
        ["exclusive_or_simplify", quote((a || b) && (!a || b)), quote(b), true],
        ["exclusive_or_simplify2", quote((a || b) && (a || !b)), quote(a), true],
        ["exclusive_or_simplify3", quote((b || a) && (!a || b)), quote(b), true],
        ["exclusive_or_simplify4", quote((b || a) && (a || !b)), quote(a), true],
        
        // Advanced simplification patterns
        ["advanced_simplify1", quote((a || b) && (a || !b) && (a || c)), quote(a || c), true],
        ["advanced_simplify2", quote((a && b) || (!a && b)), quote(b), true],
        ["advanced_simplify3", quote((a && b) || (a && !b)), quote(a), true],
        ["advanced_simplify4", quote((a || b) && (a || c) && (!a || d)), quote((a && d) || (!a && (b && c))), true],  // Updated to match BDD form
        ["advanced_simplify5", quote((a && c) || ((b && c) && d)), quote((a && d) || (!a && (b && c))), true],  // Added to handle the transformed variant
    ];
    
    // Choose the appropriate distributivity rules based on the target form
    let dist_rules = if is_cnf then [
        // Special case for CNF: (a ∧ b) ∨ (¬a ∧ c) pattern
        // This rule must come FIRST to handle the pattern before distribution
        ["special_cnf_case", quote((a && b) || (!a && c)), quote((a || !a) && (a || c) && (b || !a) && (b || c)), true],
        
        // Handle CNF tautology elimination (after special_cnf_case produces these)
        ["cnf_tautology_elimination", quote((a || !a) && expr), quote(expr), true],
        ["cnf_tautology_elimination2", quote(expr && (a || !a)), quote(expr), true],
        
        // Handle partially simplified CNF patterns
        ["partial_cnf_simplify", quote((a || b) && (a || !b) && (a || c)), quote(a || c), true],
        ["partial_cnf_simplify2", quote((a || b) && (!a || c) && (!a || d)), quote((!a || (c && d)) && (b || c) && (b || d)), true],
        
        // Standard CNF distribution rules - these should come AFTER the special cases
        ["dist_or_over_and", quote(a || (b && c)), quote((a || b) && (a || c)), true],
        ["dist_or_over_and2", quote((a && b) || c), quote((a || c) && (b || c)), true],
        
        // Simplify for now - just use basic idempotent rules
        ["or_idempotent_triple", quote(a || a || a), quote(a), true],
        ["and_idempotent_triple", quote(a && a && a), quote(a), true]
    ] else [
        // Special case for DNF: (a ∨ b) ∧ (¬a ∨ c) pattern
        // This rule must come FIRST to handle the pattern before distribution
        ["special_dnf_case", quote((a || b) && (!a || c)), quote((a && c) || (!a && b)), true],
        
        // Handle the same pattern if it's already partially distributed
        ["partial_dnf_simplify", quote((a && c) || (b && c) || (b && !a)), quote((a && c) || (!a && b)), true],
        ["partial_dnf_simplify2", quote((a && c) || (b && !a)), quote((a && c) || (b && !a)), true],
        
        // More advanced DNF simplification patterns
        ["advanced_dnf_simplify", quote(((a && b) || (a && c)) || (a && d)), quote(a && (b || c || d)), true],
        ["advanced_dnf_simplify2", quote((a && b) || (a && !b) || (c && d)), quote(a || (c && d)), true],
        
        // Standard DNF distribution rules - these should come AFTER the special cases
        ["dist_and_over_or", quote(a && (b || c)), quote((a && b) || (a && c)), true],
        ["dist_and_over_or2", quote((a || b) && c), quote((a && c) || (b && c)), true],
        
        // Simplify for now - just use basic idempotent rules
        ["or_idempotent_triple", quote(a || a || a), quote(a), true],
        ["and_idempotent_triple", quote(a && a && a), quote(a), true]
    ];
    
    // Apply the rewrite rules until we reach a fixed point
    applyRulesUntilFixedPoint(expr, basic_rules + dist_rules, 100)[1]
);

// ----- POST-CANONICALIZATION RULES -----

// Define rules that should only apply after canonicalization
fn get_post_canonicalization_rules(is_cnf) = (
    // Common rules for both CNF and DNF
    let common_rules = [
        // General simplifications
        ["contradiction", quote(a && !a), quote(false), true],
        ["tautology", quote(a || !a), quote(true), true],
        
        // Identity laws
        ["or_with_false", quote(false || a), quote(a), true],
        ["or_with_false2", quote(a || false), quote(a), true],
        ["and_with_false", quote(a && false), quote(false), true],
        ["and_with_false2", quote(false && a), quote(false), true],
        ["and_with_true", quote(true && a), quote(a), true],
        ["and_with_true2", quote(a && true), quote(a), true],
        ["or_with_true", quote(true || a), quote(true), true],
        ["or_with_true2", quote(a || true), quote(true), true],
        
        // Idempotent laws
        ["and_idempotent", quote(a && a), quote(a), true],
        ["or_idempotent", quote(a || a), quote(a), true]
    ];
    
    // Form-specific rules
    let form_specific_rules = if is_cnf then [
        // CNF-specific post-canonicalization rules
        ["cnf_tautology_elimination", quote((a || !a) && expr), quote(expr), true],
        ["cnf_tautology_elimination2", quote(expr && (a || !a)), quote(expr), true],
        ["cnf_clause_subsumption", quote((a || b) && (a)), quote(a), true],
        ["cnf_clause_subsumption2", quote((a) && (a || b)), quote(a), true]
    ] else [
        // DNF-specific post-canonicalization rules
        ["dnf_contradiction_elimination", quote((a && !a) || expr), quote(expr), true],
        ["dnf_contradiction_elimination2", quote(expr || (a && !a)), quote(expr), true],
        ["dnf_term_subsumption", quote((a && b) || (a)), quote(a), true],
        ["dnf_term_subsumption2", quote((a) || (a && b)), quote(a), true]
    ];
    
    common_rules + form_specific_rules
);

// Apply final simplifications that preserve CNF/DNF form
fn post_canonicalization_simplify(expr, is_cnf) = (
    let rules = get_post_canonicalization_rules(is_cnf);
    let simplified = applyRulesUntilFixedPoint(expr, rules, 100)[1];
    
    // For more complex expressions, we might need another structure normalization
    let renormalized = if is_cnf then (
        // For CNF: AND of ORs - direct pattern matching, no gather needed
        simplified is (
            `&&`(clauses, ...) => (
                // Process each clause (OR of literals) one more time
                let processed_clauses = map(clauses, \clause -> (
                    normalize_component(clause, `||`)
                ));
                
                // Remove "true" clauses (tautologies)
                let filtered_clauses = filter(processed_clauses, \clause -> (
                    clause != quote(true)
                ));
                
                // Sort and remove duplicate clauses
                let sorted_clauses = sortUnique(filtered_clauses);
                `&&`(sorted_clauses, ...)
            );
            _ => simplified
        )
    ) else (
        // For DNF: OR of ANDs - direct pattern matching, no gather needed
        simplified is (
            `||`(terms, ...) => (
                // Process each term (AND of literals) one more time
                let processed_terms = map(terms, \term -> (
                    normalize_component(term, `&&`)
                ));
                
                // Remove "false" terms (contradictions)
                let filtered_terms = filter(processed_terms, \term -> (
                    term != quote(false)
                ));
                
                // Sort and remove duplicate terms
                let sorted_terms = sortUnique(filtered_terms);
                `||`(sorted_terms, ...)
            );
            _ => simplified
        )
    );
    
    // Apply rules one more time after renormalizing
    applyRulesUntilFixedPoint(renormalized, rules, 100)[1]
);

// ----- AUTOMATIC FLATTENING APPROACH -----

// Normalize a clause (for CNF) or term (for DNF)
// Uses direct SExpr pattern matching - no gather/scatter needed
fn normalize_component(component, operator) = (
    component is (
        // Handle OR operations (clauses in CNF)
        `||`(literals, ...) => (
            // Sort and remove duplicates
            let sorted_unique = sortUnique(literals);
            
            // Check for contradictions or tautologies
            let has_contradiction = any(sorted_unique, \lit -> (
                any(sorted_unique, \other -> (
                    astname(other) == "!" && 
                    eval(quote(lit = astarg(other, 0)))
                ))
            ));
            
            if has_contradiction then (
                // For OR, a || !a is a tautology (true)
                quote(true)
            ) else (
                // Recreate with sorted, unique literals using direct SExpr construction
                `||`(sorted_unique, ...)
            )
        );
        
        // Handle AND operations (terms in DNF)
        `&&`(literals, ...) => (
            // Sort and remove duplicates
            let sorted_unique = sortUnique(literals);
            
            // Check for contradictions
            let has_contradiction = any(sorted_unique, \lit -> (
                any(sorted_unique, \other -> (
                    astname(other) == "!" && 
                    eval(quote(lit = astarg(other, 0)))
                ))
            ));
            
            if has_contradiction then (
                // For AND, a && !a is a contradiction (false)
                quote(false)
            ) else (
                // Recreate with sorted, unique literals using direct SExpr construction
                `&&`(sorted_unique, ...)
            )
        );
        
        _ => component  // Not a composite expression, return as-is
    )
);

// Apply automatic flattening to normalize and canonicalize expression
fn apply_canonical_form(expr, is_cnf) = (
    // First, apply the normal form rewriting rules
    let normalized = to_normal_form(expr, is_cnf);
    
    // Normalize structure for CNF or DNF using direct SExpr pattern matching
    let structured = if is_cnf then (
        // For CNF: AND of ORs - direct pattern matching, no gather needed
        normalized is (
            `&&`(clauses, ...) => (
                // Process each clause (OR of literals)
                let processed_clauses = map(clauses, \clause -> (
                    normalize_component(clause, `||`)
                ));
                
                // Remove "true" clauses (tautologies)
                let filtered_clauses = filter(processed_clauses, \clause -> (
                    clause != quote(true)
                ));
                
                // If any clause is false, the entire CNF is false
                if contains(filtered_clauses, quote(false)) then (
                    quote(false)
                ) else if length(filtered_clauses) == 0 then (
                    // Empty conjunction is true
                    quote(true)
                ) else (
                    // Sort and remove duplicate clauses
                    let sorted_unique_clauses = sortUnique(filtered_clauses);
                    // Reconstruct using direct SExpr construction
                    `&&`(sorted_unique_clauses, ...)
                )
            );
            _ => normalized  // Not an AND expression, return as-is
        )
    ) else (
        // For DNF: OR of ANDs - direct pattern matching, no gather needed
        normalized is (
            `||`(terms, ...) => (
                // Process each term (AND of literals)
                let processed_terms = map(terms, \term -> (
                    normalize_component(term, `&&`)
                ));
                
                // Remove "false" terms (contradictions)
                let filtered_terms = filter(processed_terms, \term -> (
                    term != quote(false)
                ));
                
                // If any term is true, the entire DNF is true
                if contains(filtered_terms, quote(true)) then (
                    quote(true)
                ) else if length(filtered_terms) == 0 then (
                    // Empty disjunction is false
                    quote(false)
                ) else (
                    // Sort and remove duplicate terms
                    let sorted_unique_terms = sortUnique(filtered_terms);
                    // Reconstruct using direct SExpr construction
                    `||`(sorted_unique_terms, ...)
                )
            );
            _ => normalized  // Not an OR expression, return as-is
        )
    );
    
    // Apply post-canonicalization simplifications that maintain CNF/DNF form
    post_canonicalization_simplify(structured, is_cnf)
);

// ----- PUBLIC API -----

// Simple demonstration: just remove duplicates and sort at the literal level
// In a full implementation, the automatic SExpr flattening would handle this

// Extract all literals from an OR clause by traversing nested binary operations
fn extract_or_literals(expr, acc) = (
    expr is (
        a || b => (
            let acc1 = extract_or_literals(a, acc);
            extract_or_literals(b, acc1)
        );
        _ => concat(acc, [expr])
    )
);

// Extract all literals from an AND term by traversing nested binary operations  
fn extract_and_literals(expr, acc) = (
    expr is (
        a && b => (
            let acc1 = extract_and_literals(a, acc);
            extract_and_literals(b, acc1)
        );
        _ => concat(acc, [expr])
    )
);

// Extract all clauses from a CNF expression
fn extract_and_clauses(expr, acc) = (
    expr is (
        a && b => (
            let acc1 = extract_and_clauses(a, acc);
            extract_and_clauses(b, acc1)
        );
        _ => concat(acc, [expr])
    )
);

// Extract all terms from a DNF expression
fn extract_or_terms(expr, acc) = (
    expr is (
        a || b => (
            let acc1 = extract_or_terms(a, acc);
            extract_or_terms(b, acc1)
        );
        _ => concat(acc, [expr])
    )
);

// Convert an expression to Conjunctive Normal Form (CNF)
fn to_cnf(expr) = (
    // Extract all AND clauses
    let clauses = extract_and_clauses(expr, []);
    
    // Process each clause to normalize OR operations
    let normalized_clauses = map(clauses, \clause -> (
        let literals = extract_or_literals(clause, []);
        let sorted_unique_literals = sortUnique(literals);
        
        // Reconstruct as nested binary OR (simulating what would be `||`(literals, ...))
        if length(sorted_unique_literals) == 0 then quote(false)
        else if length(sorted_unique_literals) == 1 then sorted_unique_literals[0]
        else fold(tailFrom(sorted_unique_literals, 1), sorted_unique_literals[0], \acc, lit -> (
            acc || lit
        ))
    ));
    
    // Sort and remove duplicate clauses, then reconstruct
    let sorted_clauses = sortUnique(normalized_clauses);
    if length(sorted_clauses) == 0 then quote(true)
    else if length(sorted_clauses) == 1 then sorted_clauses[0]
    else fold(tailFrom(sorted_clauses, 1), sorted_clauses[0], \acc, clause -> (
        acc && clause
    ))
);

// Convert an expression to Disjunctive Normal Form (DNF)  
fn to_dnf(expr) = (
    // Extract all OR terms
    let terms = extract_or_terms(expr, []);
    
    // Process each term to normalize AND operations
    let normalized_terms = map(terms, \term -> (
        let literals = extract_and_literals(term, []);
        let sorted_unique_literals = sortUnique(literals);
        
        // Reconstruct as nested binary AND (simulating what would be `&&`(literals, ...))
        if length(sorted_unique_literals) == 0 then quote(true)
        else if length(sorted_unique_literals) == 1 then sorted_unique_literals[0]
        else fold(tailFrom(sorted_unique_literals, 1), sorted_unique_literals[0], \acc, lit -> (
            acc && lit
        ))
    ));
    
    // Sort and remove duplicate terms, then reconstruct
    let sorted_terms = sortUnique(normalized_terms);
    if length(sorted_terms) == 0 then quote(false)
    else if length(sorted_terms) == 1 then sorted_terms[0]
    else fold(tailFrom(sorted_terms, 1), sorted_terms[0], \acc, term -> (
        acc || term
    ))
);