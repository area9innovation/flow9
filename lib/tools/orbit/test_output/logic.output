==== CNF and DNF Canonical Forms with Gather/Scatter ====
----- Test Case 1 -----
Original: ¬(a ∧ b)
CNF: (¬a ∨ ¬b)
DNF: (¬a ∨ ¬b)
----- Test Case 2 -----
Original: (a ∨ (b ∧ c))
CNF: ((a ∨ b) ∧ (a ∨ c))
DNF: (a ∨ (b ∧ c))
----- Test Case 3 -----
Original: (a ∧ (b ∨ c))
CNF: (a ∧ (b ∨ c))
DNF: ((a ∧ b) ∨ (a ∧ c))
----- Test Case 4 -----
Original: ((a ∨ b) ∧ (¬a ∨ c))
CNF: ((a ∨ b) ∧ (¬a ∨ c))
DNF: ((a ∧ c) ∨ (¬a ∧ b))
----- Test Case 5 -----
Original: (((a ∨ b) ∨ c) ∨ ((d ∧ e) ∧ f))
CNF: (((((a ∨ b) ∨ c) ∨ d) ∧ (((a ∨ b) ∨ c) ∨ e)) ∧ (((a ∨ b) ∨ c) ∨ f))
DNF: (((a ∨ b) ∨ c) ∨ ((d ∧ e) ∧ f))
----- Test Case 6 -----
Original: ((((a + b) < 2) ∨ (c > 3)) ∨ (c > 3))
CNF: ((((a + b) < 2) ∨ (c > 3)) ∨ (c > 3))
DNF: ((c > 3) ∨ ((a + b) < 2))
Result: "Done"
