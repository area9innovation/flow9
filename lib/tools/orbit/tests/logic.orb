import lib/logic;

// ----- EXAMPLE DEMONSTRATION -----

// Demonstrate CNF and DNF transformations
fn main() = (
    println("==== CNF and DNF Canonical Forms with Gather/Scatter ====");
    
    // Test expressions
    let test_expressions = [
        quote(!(a && b)),           // De Morgan's law test
        quote(a || (b && c)),       // OR over AND distribution (CNF)
        quote(a && (b || c)),       // AND over OR distribution (DNF)
        quote((a || b) && (!a || c)),  // A practical example - resolution candidate
        quote(a || b || c || (d && e && f))  // Multiple nested operators
    ];
    
    // Process each test case
    let count = 1;
    fold(test_expressions, count, \i, expr.(
        println("\n----- Test Case " + i2s(i) + " -----");
        println("Original: " + prettyOrbit(expr));
        println("CNF: " + prettyOrbit(to_cnf(expr)));
        println("DNF: " + prettyOrbit(to_dnf(expr)));
        i + 1
    ));
    
    "Done"
);

// Run the demonstration
main();