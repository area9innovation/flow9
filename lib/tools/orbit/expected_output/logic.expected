==== CNF and DNF Canonical Forms with Gather/Scatter ====
----- Test Case 1 -----
Original: ¬(a ∧ b)
Rule 'de_morgan_and' applied. ID: 7: (¬a ∨ ¬b) was ¬(a ∧ b) at 4
CNF: (¬a ∨ ¬b)
Rule 'de_morgan_and' applied. ID: 7: (¬a ∨ ¬b) was ¬(a ∧ b) at 4
DNF: (¬a ∨ ¬b)
----- Test Case 2 -----
Original: (a ∨ (b ∧ c))
Rule 'dist_or_over_and' applied. ID: 8: ((a ∨ b) ∧ (a ∨ c)) was (a ∨ (b ∧ c)) at 5
CNF: ((a ∨ b) ∧ (a ∨ c))
DNF: (a ∨ (b ∧ c))
----- Test Case 3 -----
Original: (a ∧ (b ∨ c))
CNF: (a ∧ (b ∨ c))
Rule 'dist_and_over_or' applied. ID: 8: ((a ∧ b) ∨ (a ∧ c)) was (a ∧ (b ∨ c)) at 5
DNF: ((a ∧ b) ∨ (a ∧ c))
----- Test Case 4 -----
Original: ((a ∨ b) ∧ (¬a ∨ c))
CNF: ((a ∨ b) ∧ (¬a ∨ c))
Rule 'special_dnf_case' applied. ID: 12: ((a ∧ c) ∨ (¬a ∧ b)) was ((a ∨ b) ∧ (¬a ∨ c)) at 8
DNF: ((a ∧ c) ∨ (¬a ∧ b))
----- Test Case 5 -----
Original: (((a ∨ b) ∨ c) ∨ ((d ∧ e) ∧ f))
Rule 'dist_or_over_and' applied. ID: 14: ((((a ∨ b) ∨ c) ∨ (d ∧ e)) ∧ (((a ∨ b) ∨ c) ∨ f)) was (((a ∨ b) ∨ c) ∨ ((d ∧ e) ∧ f)) at 11
Rule 'dist_or_over_and' applied. ID: 17: ((((a ∨ b) ∨ c) ∨ d) ∧ (((a ∨ b) ∨ c) ∨ e)) was (((a ∨ b) ∨ c) ∨ (d ∧ e)) at 12
CNF: (((((a ∨ b) ∨ c) ∨ d) ∧ (((a ∨ b) ∨ c) ∨ e)) ∧ (((a ∨ b) ∨ c) ∨ f))
DNF: (((a ∨ b) ∨ c) ∨ ((d ∧ e) ∧ f))
Result: "Done"
